### DISCUSSION TITLE
fcm-normalize for multiple images

### DISCUSSION THREAD

**Discussion by Al-yhuwert** [jcreinhold/intensity-normalization#78] [Q&A] [answered]
Hello,
I'm trying to normalize lots of T1 images using fcm-normalize, I used to do it like this:
`fcm-normalize -i t1_images_folder -tm wm_masks_folder -o output_folder -c t1 -tt wm -p -v`

But, the library was updated and I don't know how to work on folders like before. 

I've tried what is specified in the documents:

`find -L t1_images_folder -type f -name '*t1.nii*' -exec fcm-normalize -p -mo t1 -tt wm "{}" \;`

The thing is that I can't add the folder with my white matter masks, something like: 

`find -L t1_images_folder -type f -name '*t1.nii*' -exec fcm-normalize -p -mo t1 -tt wm -tm wm_masks_folder "{}" \;`

 I know fcm-normalize can get the wm masks internally but I already have them so I believe the algorithm can run faster having these masks ready. Also, I want to plot the histogram of all the normalized images in same plot as it used to work before the update, but now for every normalized image I get a hist.pdf plot and it is rewritten each time an image is normalized. 
 
 I hope you can help me, thanks a lot!
 
 Al.

**Comment by jcreinhold** [comment, marked as answer]
The easiest way to do this is to write a small bash script with a for loop properly formats the call to `fcm-normalize` with the corresponding image and brain mask. You can avoid using `find` in that case; [this bash reference](https://learnxinyminutes.com/docs/bash/) might be useful if you're not familiar with bash.

You can also use the Python API to write a script that does the normalization if that's easier than bash for you (see the docs for examples).

I had to remove the ability to process whole directories for individual timepoint-based normalization schemes because it was too difficult to maintain.

For the histogram, instead of using the `-p` option on `fcm-normalize`, use the `plot-histograms` CLI that's installed with the package after you normalize the set of images. Use `plot-histograms -h` to see how to call it and the options.

**Reply by Al-yhuwert** [reply]
Thanks for your answer!
Yes, I did a bash script and it's working well. 

One further question is regarding the histogram, I plotted the histogram of each image using the `-p` option and then I used `plot-histograms` as you suggested. But I think the results are quite different, is this because of the scale or am I doing something wrong?

![Screenshot from 2023-09-27 09-22-40](https://github.com/jcreinhold/intensity-normalization/assets/79812627/408a2dc8-4252-452c-b3b1-ba97ea1bd068)

**Reply by jcreinhold** [reply]
I trust the `plot-histograms` output, but I'm not sure what's happening with the individual histograms. Could you provide the bash script you used to invoke `fcm-normalize`? When you look at the images in a medical image viewer or plot the histogram of the intensities yourself, do you see the same graph?

**Reply by Al-yhuwert** [reply]
This is the script I used:
```
pth=/path/to/folder/location
cd ${pth}

for file in img/[0-9][0-9][0-9]_S_[0-9][0-9][0-9][0-9]_t1_brain_robex.nii; do
    filename=$(basename "$file")
    img_number=$(echo "$filename" | grep -oE '[0-9]{3}_S_[0-9]{4}')
    
    plot_filename="./fcm-norm/${img_number}_t1_brain_fcm_histogram.pdf"
    
    fcm-normalize -o "./fcm-norm/${img_number}_t1_robex_fcm.nii.gz" -p -mo t1 -tt wm -tm "c2/${img_number}_wm_mask.nii" "$file"
    
    # Rename the histogram file
    mv ./fcm-norm/hist.pdf "$plot_filename"
done

plot-histograms ./fcm-norm/
```

Looking at the individual histograms in `fsleyes` they look pretty similar to the output of `plot-histograms`

**Reply by jcreinhold** [reply]
Ah, this is a bit of a bug I suppose in the histogram plotting logic. Because you're passing in the tissue mask directly, it's using that as the brain mask when creating the histogram. So the individual histogram represents the WM of the image after normalization—since it's T1-w it seems correct to have the gap between 0 and the start of the foreground. I assume you're using the brain mask when calling `plot-histograms` which is why that looks normal.

**Reply by jcreinhold** [reply]
Opened this bug as an issue: #79. Not sure if I'll bother to fix it or if there is a great alternative solution in the first place, but FYI.
